core/src/main/java/org/apache/iceberg/encryption/Ciphers.java

```diff
@@ -55,10 +71,15 @@ public AesGcmEncryptor(byte[] keyBytes) {
       this.randomGenerator = new SecureRandom();
     }
 
-    public byte[] encrypt(byte[] plainText, byte[] aad) {
+    public byte[] encrypt(byte[] plaintext, byte[] aad) {
+      return encrypt(plaintext, 0, plaintext.length, aad);
+    }
+
+    public byte[] encrypt(byte[] plaintext, int plaintextOffset, int plaintextLength, byte[] aad) {
+      Preconditions.checkArgument(plaintextLength > 0, "Wrong plaintextLength " + plaintextLength);
       byte[] nonce = new byte[NONCE_LENGTH];
       randomGenerator.nextBytes(nonce);
-      int cipherTextLength = NONCE_LENGTH + plainText.length + GCM_TAG_LENGTH;
+      int cipherTextLength = NONCE_LENGTH + plaintextLength + GCM_TAG_LENGTH;
```

Comment:
Is this buffer allocation necessary? Why can't we pass this in and reuse it for every block? It is passed into the `doFinal` call so I think we could have this return the number of bytes that were encrypted and reuse a buffer from the caller.

Reply:
Done for simplicity, since manifest encryption won't benefit much from this optimization. 
But we can add now, for Avro data encryption in the future.
Btw, the same optimization can be performed for decryption.

Reply:
Added to both encryption and decryption in the new commit.